CSharp/Library/Microsoft.Bot.Connector.Shared/MicrosoftAppCredentials.cs

```diff
@@ -19,6 +19,19 @@
 
 namespace Microsoft.Bot.Connector
 {
+
+    public static class TimeSpanExtensions
+    {
+        public static TimeSpan WithRandomNoise(this TimeSpan delay, double multiplier = 0.1)
+        {
+            // Generate an uniform distribution between zero and 10% of the proposed delay and add it as 
+            // random noise. The reason for this is that if there are multiple threads about to retry
+            // at the same time, it can overload the server again and trigger throttling again.
+            // By adding a bit of random noise, we distribute requests a across time.
+            return delay + TimeSpan.FromMilliseconds(new Random().NextDouble() * delay.TotalMilliseconds * 0.1);

```

Comment:
At the risk of being picky, creating a new Random() each time is expensive. Random is probably best as a static variable, and the functions like NextDouble() are thread safe.

Reply:
Normally the phrase here would be "Jitter"

Reply:
This is perfectly valid, not pick at all!

Reply:
I also moved it to Retry.cs since it is used there.